### PR TITLE
ITextureProvider updates

### PR DESCRIPTION
**Pull request by Soreepeong**
This PR will obsolete all texture loading functions from `UiBuilder` and put all of them to `ITextureProvider`. Functions that do not complete immediately (due to filesystem access and such) are only given async versions. All async versions are also given option to pass a `CancellationToken`.

At a time, at most `min(64, Environment.ProcessorCount)` textures will be loaded concurrently. Async texture requests are given priorities, and will be processed in order of request; immediate texture requests will be processed in order of most recently requested.

This change makes IconBrowserWidget scroll do not stutter.
image

## Deprecated
Functions returning textures had ambiguous lifetime, especially when one was trying to store ImGuiHandle. Such functions have been split into two; one for immediate use for rendering, and another for ownership transferral.
* `ITextureProvider.GetIconPath(uint, ...)`: use `TryGetIconPath(GameIconLookup, out string)` or `GetIconPath(GameIconLookup)` instead.
* `ITextureProvider.GetIcon`: use `GetFromGameIcon(...)` instead.
* `ITextureProvider.GetTextureFromGame`: use `GetFromGame(...)` instead.
* `ITextureProvider.GetTextureFromFile`: use `GetFromFile` instead.
* `UiBuilder.LoadImage...`: use functions in `ITextureProvider`.

### CODE REVIEW

**Comment by grittyfrog** [comment]
Do we know what the timeline might look like for merging this PR? 

I'm looking to submit an Icon Picker widget to Dalamud, and the changes in this PR simplify it a lot since I don't need to maintain my own texture cache. 

[This is what I'm planning to submit](https://github.com/grittyfrog/MacroMate/blob/master/MacroMate/Extensions/Dalamud/Interface/ImGuiIconPicker/IconPickerDialog.cs):

[iconpicker.webm](https://github.com/goatcorp/Dalamud/assets/148605153/66113496-4f77-4ebd-8797-1062064675be)

And the changes here I think should let me get rid of [TextureCache](https://github.com/grittyfrog/MacroMate/blob/54ab8071bbc3b6cbf201caacb357a646286301d4/MacroMate/Extensions/Dalamud/Texture/TextureCache.cs)

**Comment by goaaats** [comment]
Should we merge this into apiX directly instead?

**Comment by Soreepeong** [comment]
Changed target branch, and merged apiX. Not sure why GH is displaying conflicting files

**Comment by goaaats** [comment]
Thanks! Will start giving it a look-through.

File: Dalamud/Interface/Textures/Internal/TextureManager.Wic.cs [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+    {
+        try
+        {
+            if (wrap is null)
+                throw new NullReferenceException($"{nameof(wrap)} cannot be null.");
+            if (path is null)
+                throw new NullReferenceException($"{nameof(path)} cannot be null.");
+
+            using var wrapAux = new WrapAux(wrap, true);
+            var pathTemp = $"{path}.{GetCurrentThreadId():X08}{Environment.TickCount64:X16}.tmp";
```
Comment (goaaats) [inline Dalamud/Interface/Textures/Internal/TextureManager.Wic.cs:93]: Is this safe? In my experience this kinda code always blows up. Maybe we need something like this? [GetTempPath](https://github.com/goatcorp/FFXIVQuickLauncher/blob/6e0829a973b26db3aa95a9e60b96d73ff222a1ae/src/XIVLauncher.Common/Util/PlatformHelpers.cs#L34)

File: Dalamud/Interface/Textures/Internal/TextureManager.Drawer.cs [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+internal sealed partial class TextureManager
+{
+    private SimpleDrawerImpl? simpleDrawer;
+
+    /// <summary>A class for drawing simple stuff.</summary>
+    [SuppressMessage(
+        "StyleCop.CSharp.LayoutRules",
+        "SA1519:Braces should not be omitted from multi-line child statement",
+        Justification = "Multiple fixed blocks")]
+    internal sealed unsafe class SimpleDrawerImpl : IDisposable
```
Comment (goaaats) [inline Dalamud/Interface/Textures/Internal/TextureManager.Drawer.cs:25]: If I understand this code and the usage correctly, this is only used for "copying" a texture by drawing it to another? I don't know a lot about DirectX, why is this necessary?

**Review by goaaats — CHANGES_REQUESTED** [review]
Looks great, thank you - and sorry for the huge delay on the review! Left a few questions I had, other than that, I think this is good to go from an API-perspective.

Note that I'm not really in a great position to review the DirectX API usage in-depth so I'll trust you on this and hope that you're around to troubleshoot if there were any issues with it.

File: Dalamud/Interface/Internal/InterfaceManager.cs [hunk 3]
```diff
@@ -697,24 +656,35 @@ private IntPtr PresentDetour(IntPtr swapChain, uint syncInterval, uint presentFl
             return this.presentHook!.Original(swapChain, syncInterval, presentFlags);
         }
 
+        this.CumulativePresentCalls++;
+        this.IsInPresent = true;
```
[the remark this answers is not in the document]
Comment (Soreepeong) [reply Dalamud/Interface/Internal/InterfaceManager.cs:660]: These do not need to be thread safe. Only written from the main thread, and for all other reader methods are expected to know that these variables will be touched from the main thread.

File: Dalamud/Interface/Internal/Windows/Data/Widgets/IconBrowserWidget.cs [hunk 4]
```diff
@@ -1,12 +1,11 @@
 ﻿using System.Collections.Generic;
-using System.Drawing;
-using System.Linq;
 using System.Numerics;
+using System.Threading.Tasks;
```
[the remark this answers is not in the document]
Comment (Soreepeong) [reply Dalamud/Interface/Internal/Windows/Data/Widgets/IconBrowserWidget.cs:3]: I think someone else was intending to make a full-featured icon browser but was waiting for this PR, so I think it's best to leave it internal for now so that that other PR won't have any compatibility constraints.

File: Dalamud/Interface/Textures/ISharedImmediateTexture.cs [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    /// <summary>Creates a new instance of <see cref="IDalamudTextureWrap"/> holding a new reference to this texture.
+    /// The returned texture is guaranteed to be available until <see cref="IDisposable.Dispose"/> is called.</summary>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A <see cref="Task{TResult}"/> containing the loaded texture on success.</returns>
+    /// <remarks>
+    /// <see cref="IDisposable.Dispose"/> must be called on the resulting instance of <see cref="IDalamudTextureWrap"/>
+    /// from the returned <see cref="Task{TResult}"/> after use. Consider using
+    /// <see cref="DisposeSafety.ToContentDisposedTask{T}"/> to dispose the result automatically according to the state
+    /// of the task.</remarks>
+    Task<IDalamudTextureWrap> RentAsync(CancellationToken cancellationToken = default);
```
[the remark this answers is not in the document]
Comment (Soreepeong) [reply Dalamud/Interface/Textures/ISharedImmediateTexture.cs:69]: This function is for plugins that exactly knows when a texture can be (pre)loaded and until when should it be kept in memory, or for passing the contained hardware resource to APIs that are not managed by Dalamud. Also, while it states that it must be disposed, it does have a finalizer that will decrease the reference counter.

File: Dalamud/Interface/Internal/Windows/Data/Widgets/IconBrowserWidget.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MidoriKami) [reply Dalamud/Interface/Internal/Windows/Data/Widgets/IconBrowserWidget.cs:3]: This one was intended to be a more "raw" icon browser that shows all the icons without any filtering or categorization. QoL's icon browser suffered from over filtering when new icons were added to the game. The browser that someone was wanting to make categorizes them manually.

File: Dalamud/Interface/Textures/Internal/SharedImmediateTextures/FileSystemSharedImmediateTexture.cs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+    /// <inheritdoc/>
+    public override string ToString() =>
+        $"{nameof(FileSystemSharedImmediateTexture)}#{this.InstanceIdForDebug}({this.path})";
+
+    /// <inheritdoc/>
+    protected override async Task<IDalamudTextureWrap> CreateTextureAsync(CancellationToken cancellationToken)
+    {
+        var tm = await Service<TextureManager>.GetAsync();
+        var wrap = await tm.NoThrottleCreateFromFileAsync(this.path, cancellationToken);
```
[the remark this answers is not in the document]
Comment (Soreepeong) [reply Dalamud/Interface/Textures/Internal/SharedImmediateTextures/FileSystemSharedImmediateTexture.cs:32]: SharedImmediateTexture.CreateTextureAsync is called from scheduler; see SharedImmediateTexture.LoadUnderlyingWrap. I'll modify CTA to include a comment on why this is not throttled.

File: Dalamud/Interface/Textures/Internal/TextureManager.Wic.cs (hunk 1 above)
Comment (Soreepeong) [reply Dalamud/Interface/Textures/Internal/TextureManager.Wic.cs:93]: Creating a file in a temporary directory may end up creating a file in a different drive, so that move/replace operation cannot happen without copying. While I'm expecting this to be safe (thread-unique, effectively monotonic counter) to be safe and to offer a single point that needs to be fixed in case of errors, I'll add a utility function for making a temporary file name that should be able to be used for file replace operations.

File: Dalamud/Interface/Textures/Internal/TextureManager.Drawer.cs (hunk 2 above)
Comment (Soreepeong) [reply Dalamud/Interface/Textures/Internal/TextureManager.Drawer.cs:25]: `.Draw` should be taken as a hardware-accelerated pixel format conversion. `.StripAlpha` that follows will touch all pixels so that all alpha values become 1, which cannot be done with the functions exposed from ImGui, and would be noticeably more time consuming if the process is done by taking the pixels off GPU, process it with CPU, and upload to GPU again.

**Comment by Soreepeong** [comment]
Addressed the review comments, and merged apiX.

**Review by goaaats — APPROVED** [review]
